### ISSUE TITLE
Youtube links can no longer be opened with IINA

### ISSUE THREAD

**Issue by Mockedarche** [iina/iina#4223]
Hello, I always use IINA when watching YouTube videos. Starting yesterday 2/16/23 whenever I open a YouTube link in IINA I get "can't open file or stream!". Put simply YouTube links no longer appear to work in IINA and this is the first time I've ever had IINA have an issue. This is after around 2-3 years of using IINA for this.
Checked open issues and none appear to mention Youtube.

macOS 13.2
IINA 1.3.1

**Expected behavior:**
IINA opens the link and buffers for a few seconds and then the video is playable. 

**Actual behavior:**
IINA attempts to open the link and fails.

**How often does this happen?**
Every single YouTube video fails to play. Videos play correctly when it's local and off my SMB share. Appears to solely be a problem when playing a YouTube link. 

Let me know if there is anything I can do to help!

**Comment by marcopolo4245** [comment]
Confirmed ,  Adding the open in also does not work, in firefox , 13.2.1

**Comment by casualbugs** [comment]
Same problem on Brave. Perhaps YouTube changed something?

**Comment by marcopolo4245** [comment]
Perhaps, 
My stop working after the Mac os sercurity update 13.2.1, As far I am concern there was webkit security breach , that is probably the reason. 

 I try to use the youtube-dl  latest one in custom folder  in IINA but it did not work eigher .

**Comment by low-batt** [comment]
For streaming IINA 1.3.1 embeds a copy of [yt-dlp](https://github.com/yt-dlp/yt-dlp). It appears YouTube changed something and IINA version of `yt-dlp` is too old to work with YouTube now.

I just tested version 2023.02.17 of `yt-dlp` running under macOS 13.2.1 and it worked. You can download that version and in IINA's settings under the `Network` tab change IINA to use that version.

**Comment by Mockedarche** [comment]
> For streaming IINA 1.3.1 embeds a copy of [yt-dlp](https://github.com/yt-dlp/yt-dlp). It appears YouTube changed something and IINA version of `yt-dlp` is too old to work with YouTube now.
> 
> I just tested version 2023.02.17 of `yt-dlp` running under macOS 13.2.1 and it worked. You can download that version and in IINA's settings under the `Network` tab change IINA to use that version.

WORKAROUND low-batt's suggestion works for me and now IINA can open YouTube videos.

Worked for me!!! many thanks. For those wondering yt-dlp can be installed via home-brew. Then do brew info yt-dlp and then copy the given address and add /bin to it. Downsides is whenever yt-dlp updates the location will change. But for now this works and it seems to actually load stuff a smidge faster than previously. 

NOTE im keeping this open since IINA will need to update whatever version of yt-dlp is used to work around what appears to be a YouTube change.

**Comment by MetalGooseSolid** [comment]
> Worked for me!!! many thanks. For those wondering yt-dlp can be installed via home-brew. Then do brew info yt-dlp and then copy the given address and add /bin to it. Downsides is whenever yt-dlp updates the location will change. But for now this works and it seems to actually load stuff a smidge faster than previously.

Homebrew manages binaries internally under `Cellar`. You are intended to reference binaries via symlink from the `bin` directory under `/usr/local/` on intel or `/opt/homebrew/` on arm. This may be a little bit complicated, but it is at least mentioned near the top of the [brew homepage](https://brew.sh/), right under `What Does Homebrew Do?` Doing it this way will keep references intact across brew updates. You can verify the intended reference location to the binary by typing `which yt-dlp` into the terminal eg:

```sh
➜ which yt-dlp
/opt/homebrew/bin/yt-dlp
```

IINA wants the directory path, so you would type `/opt/homebrew/bin/` into the `custom youtube-dl path` setting of IINA in this case.

**Comment by FlorianWendelborn** [comment]
Since there’s still no easy comment that summarizes it all, these are the steps for the workaround

1. `brew install yt-dlp`
2. `which yt-dlp` and remove the `/yt-dlp` in the end
  _(in my case it was `/usr/local/bin/yt-dlp` => `/usr/local/bin`)_
3. In `IINA -> Settings -> Network -> youtube-dl` add the folder from above (e.g. `/usr/local/bin`)
![image](https://user-images.githubusercontent.com/1133858/220641017-95be77ac-27c0-4f53-9a27-5ea9c6aadeed.png)
4. Restart IINA

Keywords: workaround, fix, closed, solved, tldr

**Comment by vsheg** [comment]
I've used custom `yt-dlp` installed via pip. IINA (1.3.1 Build 133, macOS 13.2.1 22D68) has stopped working with it. I updated my `yt-dlp` installation, then disabled custom `yt-dlp`, no success both. Uninstalled and installed `yt-dlp` from brew, changed custom path in settings, reloaded, and IINA still can't open any youtube stream

**Comment by low-batt** [comment]
@FlorianWendelborn Nice summary!

I've heard from a user who is still having trouble. Not clear what went wrong, so here is some background information on the issue and instructions on how to check that you have successfully installed the working version of `yt-dlp`…

The origin of `yt-dlp` releases is the [yt-dlp](https://github.com/yt-dlp/yt-dlp) GitHub project. The latest release is found under [releases](https://github.com/yt-dlp/yt-dlp/releases) and can be downloaded directly from GitHub. For macOS it is desirable to use `yt-dlp_macos` , a universal macOS standalone executable that bundles dependencies such as `Python`, since Apple no longer includes `Python` in macOS.

However it is usually easier to use a package manager such as [Homebrew](https://brew.sh/) to manage installation of open source packages. Using `brew` to install `yt-dlp` will also install dependencies such as [FFmpeg](https://ffmpeg.org/) . **But** `brew upgrade yt-dlp` may **not** install the latest version of `yt-dlp`. At least that is what I encountered when I tried to upgrade to the fixed verison when this first happened.

The version `brew` installs for a package is controlled by a [Homebrew Formula](https://formulae.brew.sh/). The [yt-dlp brew formula](https://formulae.brew.sh/formula/yt-dlp) is currently configured to install:
> Current versions:
> stable 	✅ 	2023.2.17

That is the updated version that works. I just tried `brew upgrade yt-dlp` again and this time it updated `yt-dlp` to the fixed version.

Checking the GitHub history for the Ruby formula source file, [yt-dlp.rb](https://github.com/Homebrew/homebrew-core/commits/874a163a29abc62cc5eca7a90c35bc2b9d86e8e5/Formula/yt-dlp.rb), shows it was updated on Feb 17, 2023, the same day of the `yt-dlp` update release. I am confused as to why I initially had trouble getting `brew` to update to the fixed version.

What I can say for certain is that it is possible that `brew upgrade` may not upgrade a package to the latest release. That can been seen right now with `mpv`. The [mpv brew formula](https://formulae.brew.sh/formula/mpv) contains this:
> Current versions:
> stable 	✅ 	0.35.0

Even though currently the latest [mpv release](https://github.com/mpv-player/mpv/releases) is v0.35.1.

Therefore it is important to find out what version of `yt-dlp` is installed on your Mac and compare that to the latest release under [releases](https://github.com/yt-dlp/yt-dlp/releases) in the GitHub `yt-dlp` project. To check the version of `yt-dlp` execute the command `yt-dlp --version` as shown in the following [Terminal](https://support.apple.com/guide/terminal/welcome/mac) session:
 ```bash
low-batt@gag ~$ which yt-dlp
/opt/homebrew/bin/yt-dlp
low-batt@gag ~$ yt-dlp --version
2023.02.17
low-batt@gag ~$ 
```
 
 Version `2023.02.17` is the new version that works with YouTube.
 
Notice `yt-dlp` is in `/opt/homebrew/bin`, not `/usr/local/bin` as shown in the screen shot from @FlorianWendelborn above. This has to do with Homebrew. From the Homebrew [FAQ](https://docs.brew.sh/FAQ):
> The default prefix is:
> - /usr/local for macOS on Intel,
>  - /opt/homebrew for macOS on Apple Silicon/ARM

I'm testing on a MacBookPro18,2 with the M1 chip.

Once you have confirmed the working version of `yt-dlp` is installed on your Mac then you must follow these instructions:
- Start IINA
- Click on  `Settings…` under the `IINA` menu
- The settings panel appears
- On the left side of the panel click on `Network`
- For the `Custom youtube-dl path` setting specify the full pathname to the directory containing `yt-dlp` 
- **Restart IINA** to activate the new settings

I have confirmed IINA works with these settings and the latest `yt-dlp` installed by `brew`.

**Comment by 10vorne** [comment]
> Since there’s still no easy comment that summarizes it all, these are the steps for the workaround
> 
> 1. `brew install yt-dlp`
> 2. `which yt-dlp` and remove the `/yt-dlp` in the end
>    _(in my case it was `/usr/local/bin/yt-dlp` => `/usr/local/bin`)_
> 3. In `IINA -> Settings -> Network -> youtube-dl` add the folder from above (e.g. `/usr/local/bin`)
>    ![image](https://user-images.githubusercontent.com/1133858/220641017-95be77ac-27c0-4f53-9a27-5ea9c6aadeed.png)
> 4. Restart IINA
> 
> Keywords: workaround, fix, closed, solved, tldr

I tried it the way you summarized....

`% brew install yt-dlp_macos 
Warning: No available formula with the name "yt-dlp_macos".
==> Searching for similarly named formulae and casks...
Error: No formulae or casks found for yt-dlp_macos.`

I'm no Bash-Expert. What did I do wrong?

**Comment by vsheg** [comment]
@10vorne `brew install yt-dlp`

**Comment by 10vorne** [comment]
> @10vorne `brew install yt-dlp`

I first downloaded yt-dlp from here https://github.com/yt-dlp/yt-dlp/releases and tried to install it via brew.

Now I tried...

% brew install yt yt-dlp
Warning: No available formula with the name "yt".
==> Searching for similarly named formulae and casks...
==> Formulae
app-engine-python          kytea                      python-typing-extensions
boost-python3              libcython                  python-yq
bpython                    micropython                python@3.10
bpytop                     mytop                      python@3.11
byteman                    proxytunnel                python@3.7
cherrytree                 ptpython                   python@3.8
crytic-compile             python-build               python@3.9
cython                     python-gdbm@3.11           pythran
easy-tag                   python-launcher            pytorch
gomodifytags               python-lsp-server          pytouhou
gst-python                 python-markdown            reorder-python-imports
inotify-tools              python-tabulate            spotify-tui
ipython                    python-tk@3.10             wxpython
jupyterlab                 python-tk@3.11             yt-dlp
jython                     python-tk@3.9

To install app-engine-python, run:
  brew install app-engine-python

==> Casks
ananas-analytics-desktop-edition         jupyter-notebook-viewer
anytype                                  jupyterlab
awips-python                             luyten
cloudytabs                               malwarebytes
copytranslator                           milkytracker
egnyte                                   mysql-connector-python
google-analytics-opt-out                 tigerjython
jclasslib-bytecode-viewer                yt-music
jupyter-notebook-ql                      ytmdesktop-youtube-music

To install ananas-analytics-desktop-edition, run:
  brew install --cask ananas-analytics-desktop-edition

**Comment by low-batt** [comment]
@10vorne The `yt-dlp_macos` refers to the name of the file download to pick when obtaining `yt-dlp` directly from the GitHub project,

Sorry for the confusion.

As shown above by @vsheg, when using `brew` just `yt-dlp` must be used.

**Comment by 10vorne** [comment]
> @10vorne `brew install yt-dlp`

Ok, that worked.... I first tried to install the downloaded version from https://github.com/yt-dlp/yt-dlp/releases

IIna is running fine with YT now.

But I'm not sure why the first attempt didn't work, is the downloaded version not meant to be installed via brew?

**Comment by low-batt** [comment]
Correct. Homebrew manages downloading packages and dependencies itself. It doesn't provide a way to install a package that you have downloaded.

Because `brew` handles dependencies it is preferred over installing by hand from the [yt-dlp](https://github.com/yt-dlp/yt-dlp) GitHub project. If you download `yt-dlp` directly from the project you are responsible for renaming the file to `yt-dlp`. You get to decide what directory the file goes in. And you must install `FFmpeg`.

I can see now how confusing my post was. Sorry about that.

What really is needed is a FAQ entry about this. Been on my list for a long time to create one. Just been too busy.

**Comment by 10vorne** [comment]
Thanks for the clarification :) I'm happy it works now....

**Comment by MetalGooseSolid** [comment]
> For macOS it is desirable to use `yt-dlp_macos` , a universal macOS standalone executable that bundles dependencies such as `Python`, since Apple no longer includes `Python` in macOS.

Just wanted to point out that yt-dlp standalone does bundle its own dependencies like Python3, but Apple does ship Python3 with macOS. Python3 is referred to as `python3` rather than `python`, but this is also true with homebrew, and with the standalone yt-dlp bundle.

**Comment by marcopolo4245** [comment]
How save is that homebrew, before I install it , anywhere.  let say on uni computer? Iina it self it a dmg file, quick and easy to remove. Hard to prove anything with homebrew do not know.

**Comment by along02** [comment]
I followed all the steps but some YouTube videos still won't open. Additionally, none of the audio is working for the videos. The sound bar is blacked out (no audio tracks).  I'm on M1 and am on yt-dlp 2023.03.04.

**Comment by tireyfi1** [comment]
I installed the latest release of yt-dlp (2023.03.04) through the standalone binary and pointed to it in IINA's preferences and while videos do seem to play again, I'm experiencing the same problem as @along02 where the audio stream seems to be missing and doesn't playback. Running macOS 10.15.7 on x86-64.

**Comment by low-batt** [comment]
Unfortunately the failures with `yt-dlp` v2023.03.04 are a different problem than what is being discussed in this issue, which is about the IINA copy of `yt-dlp` being out of date. I have opened issue #4253 for the missing audio problem.

**Comment by low-batt** [comment]
I have posted the workaround required to get `yt-dlp` v2023.03.04 working in that issue.

**Comment by Fl00der1000** [comment]
> @FlorianWendelborn Nice summary!
> 
> I've heard from a user who is still having trouble. Not clear what went wrong, so here is some background information on the issue and instructions on how to check that you have successfully installed the working version of `yt-dlp`…
> 
> The origin of `yt-dlp` releases is the [yt-dlp](https://github.com/yt-dlp/yt-dlp) GitHub project. The latest release is found under [releases](https://github.com/yt-dlp/yt-dlp/releases) and can be downloaded directly from GitHub. For macOS it is desirable to use `yt-dlp_macos` , a universal macOS standalone executable that bundles dependencies such as `Python`, since Apple no longer includes `Python` in macOS.
> 
> However it is usually easier to use a package manager such as [Homebrew](https://brew.sh/) to manage installation of open source packages. Using `brew` to install `yt-dlp` will also install dependencies such as [FFmpeg](https://ffmpeg.org/) . **But** `brew upgrade yt-dlp` may **not** install the latest version of `yt-dlp`. At least that is what I encountered when I tried to upgrade to the fixed verison when this first happened.
> 
> The version `brew` installs for a package is controlled by a [Homebrew Formula](https://formulae.brew.sh/). The [yt-dlp brew formula](https://formulae.brew.sh/formula/yt-dlp) is currently configured to install:
> 
> > Current versions:
> > stable 	✅ 	2023.2.17
> 
> That is the updated version that works. I just tried `brew upgrade yt-dlp` again and this time it updated `yt-dlp` to the fixed version.
> 
> Checking the GitHub history for the Ruby formula source file, [yt-dlp.rb](https://github.com/Homebrew/homebrew-core/commits/874a163a29abc62cc5eca7a90c35bc2b9d86e8e5/Formula/yt-dlp.rb), shows it was updated on Feb 17, 2023, the same day of the `yt-dlp` update release. I am confused as to why I initially had trouble getting `brew` to update to the fixed version.
> 
> What I can say for certain is that it is possible that `brew upgrade` may not upgrade a package to the latest release. That can been seen right now with `mpv`. The [mpv brew formula](https://formulae.brew.sh/formula/mpv) contains this:
> 
> > Current versions:
> > stable 	✅ 	0.35.0
> 
> Even though currently the latest [mpv release](https://github.com/mpv-player/mpv/releases) is v0.35.1.
> 
> Therefore it is important to find out what version of `yt-dlp` is installed on your Mac and compare that to the latest release under [releases](https://github.com/yt-dlp/yt-dlp/releases) in the GitHub `yt-dlp` project. To check the version of `yt-dlp` execute the command `yt-dlp --version` as shown in the following [Terminal](https://support.apple.com/guide/terminal/welcome/mac) session:
> 
> ```shell
> low-batt@gag ~$ which yt-dlp
> /opt/homebrew/bin/yt-dlp
> low-batt@gag ~$ yt-dlp --version
> 2023.02.17
> low-batt@gag ~$ 
> ```
> 
> Version `2023.02.17` is the new version that works with YouTube.
> 
> Notice `yt-dlp` is in `/opt/homebrew/bin`, not `/usr/local/bin` as shown in the screen shot from @FlorianWendelborn above. This has to do with Homebrew. From the Homebrew [FAQ](https://docs.brew.sh/FAQ):
> 
> > The default prefix is:
> > 
> > * /usr/local for macOS on Intel,
> > * /opt/homebrew for macOS on Apple Silicon/ARM
> 
> I'm testing on a MacBookPro18,2 with the M1 chip.
> 
> Once you have confirmed the working version of `yt-dlp` is installed on your Mac then you must follow these instructions:
> 
> * Start IINA
> * Click on  `Settings…` under the `IINA` menu
> * The settings panel appears
> * On the left side of the panel click on `Network`
> * For the `Custom youtube-dl path` setting specify the full pathname to the directory containing `yt-dlp`
> * **Restart IINA** to activate the new settings
> 
> I have confirmed IINA works with these settings and the latest `yt-dlp` installed by `brew`.

I followed the guide all the way, and I still get the 'Cannot open file or stream!' error. Using IINA version 1.3.1 and yt-dlp version 2023.03.04.

**Comment by low-batt** [comment]
@Fl00der1000 Sorry to hear you are still having trouble.

The workaround specified in this issue configures IINA to use an external version of `yt-dlp` instead of the older version embedded inside IINA. That was working for a too brief moment with `yt-dlp` 2023.02.17, but the release of `yt-dlp` 2023.03.04 introduced a new problem.

See issue #4253 for details on the new problem and a second workaround that is now required in addition to the workaround specified in this issue for  `yt-dlp` 2023.03.04 to work properly. Although the title of issue #4253 is about loss of audio it can also trigger the `Cannot open file or stream!` error. See the issue for details.

**Comment by foldfree** [comment]
the fork [iina+](https://github.com/CarterLi/iina) does not embed yt-dlp anymore and instead load whatever yt-dlp is installed on the computer.
It is working well on my machine so I wonder if you could backport the changes made from that fork?

**Comment by niclake** [comment]
> Since there’s still no easy comment that summarizes it all, these are the steps for the workaround
> 
> 1. `brew install yt-dlp`
> 2. `which yt-dlp` and remove the `/yt-dlp` in the end
>    _(in my case it was `/usr/local/bin/yt-dlp` => `/usr/local/bin`)_
> 3. In `IINA -> Settings -> Network -> youtube-dl` add the folder from above (e.g. `/usr/local/bin`)
>    ![image](https://user-images.githubusercontent.com/1133858/220641017-95be77ac-27c0-4f53-9a27-5ea9c6aadeed.png)
> 4. Restart IINA
> 
> Keywords: workaround, fix, closed, solved, tldr

Worked like a charm. Thank you so much!

**Comment by ShlomoCode** [comment]
@low-batt I wonder why iina doesn't switch itself to yt-dlp to make the streaming feature work out of the box?

**Comment by low-batt** [comment]
Not sure what you mean. Do you mean if IINA finds a version of yt-dlp on 
the path automatically prefer that over the version of yt-dlp embedded 
inside the IINA app?

Something like that?

**Comment by ShlomoCode** [comment]
@low-batt The streaming didn't work for me for YouTube/other sites (only for a direct link to an mp4 file), and in the software it's called "youtube-dl", so I thought the iina uses youtube-dl which doesn't have a stable version, only nightly builds from 2021, and not in yt- dlp, and instructs people here to manually change their computer to yt-dlp.

After that I saw in the source code that it was just a nickname, and in fact you do use yt-dlp so I deleted the comment

I just don't understand why it didn't work for me with the latest version of iina, and after I set a custom path for the yt-dlp brew install it started working

**Comment by low-batt** [comment]
Yes, IINA switched from youtube-dl to yt-dlp, but did not change the 
text found in settings. Definitely confusing. Sorry about that.

yt-dlp is reading and parsing web site pages designed for humans. Web 
sites are changing all the time. This can break yt-dlp. So yt-dlp needs 
to change frequently to continue working. That means the yt-dlp 
executable embedded inside IINA quickly becomes out of date and may stop 
working for some sites. It may be the embedded yt-dlp is already too out 
of date to work.

I believe there are plans to address this in some way, but as others are 
handling this I don't know exactly how they plan to improve this.

**Comment by ksdavidc** [comment]
just a followup on this, for others:

I wanted to install youtube-dl, (which is what IIna expects, but youtube-dl has been replaced by yt-dlp) on a sonoma intel MAC, without needing to install a package manager like brew or macports (since I don't regularly use these and don't want to install a whole bunch of extra stuff)

What seems to have worked for me, though I am not sure it is the best or recommended way:

- I downloaded the binary [yt-dlp_macos](https://github.com/yt-dlp/yt-dlp/releases/latest/download/yt-dlp_macos) from the yt-dlp website to my hard drive
- moved it (manually using finder, or otherwise) to `/usr/local/bin`. I was prompted to enter password.
- make it executable `sudo chmod a+rx /usr/local/bin/yt-dlp_macos`
- checked that it was working: `yt-dlp_macos --version`
- created an alias like this: `sudo ln -s /usr/local/bin/yt-dlp_macos /usr/local/bin/youtube-dl` 
- (Actually, at first I just renamed it, but I think this is bad practice, then I undid)
- Went to IIna settings network tab and put `/usr/local/bin/` as the location for youtube-dl (guys, you really should change that!!!) 

Though this works, it appears from what I read above that IIna is actually looking for yt-dlp, so, for example, I could have (should have?) made an alias of yt-dlp_macos to yt-dlp instead. I hope the settings page is corrected, expecially as the ?-help link sends you to youtube-dl not yt-dlp, which led to a long rabbit-hole before I figured it all out.

I have not thoroughly tested this, but it is working for now.

**Comment by low-batt** [comment]
I've not tested it, but I believe IINA looks for both `youtube-dl` and `yt-dlp`. IINA 1.3.5 embeds `yt-dlp`. Of course that version of `yt-dlp` is too old at this point.

IINA has been working on a plugin system. One reason for the plugin system is to move some features from a built-in implementation to a plugin implementation. This transition is occurring now.

IINA 1.4.0-beta1 shipped with several official IINA plugins installed. One of them is the Online Media plugin. One of the features of this plugin is that by default it has its own copy of `yt-dlp` and makes it easy to update it.

On using `yt-dlp_macos`…

From the issue [Online media plugin takes a long time to show yt-dlp version](https://github.com/iina/plugin-online-media/issues/3):
> `yt-dlp_macos` is a pyinstaller-bundled "onefile" executable, which means every time you execute it, it first needs to unpack its contents (~30-50MB of shared libraries and script files) into a temporary directory before yt-dlp itself is executed. These temp files are cleaned up once the executable has finished, so the security scanning would be invoked on every run, not just the first run.

If you notice `yt-dlp --version` takes a moment to respond, it is due to the time it takes to unpack the bundle before it can be run.

The Online Media plugin was updated to not use  `yt-dlp_macos` to avoid this performance penalty.

That `?` button should have been updated along with the text that still says `youtube-dl`. So much to do…

**Comment by ksdavidc** [comment]
@low-batt Thank you for this very useful explainer. It was quite clear and thorough, especially for someone with only basic unix knowledge who isn't really inclined to adding too much or mucking about with the internals of my system.

It is quite sluggish, as you suggested. I tried to use yt-dlt originally, but when I tested it I kept getting "can't run file or stream" and testing gave messages like "/usr/local/bin/yt-dlt: line 1: Not: command not found" message, so I went with the other binary. I figured it was something wrong with my system.  

I think I'll wait until IIna gets updated before trying to get it working more smoothly, though. I am using IIna mostly for local videos anyhoo. 


UPDATE: I gave it one last try, having in the meantime reluctantly installed xcode, because I installed Open Video Downloader, which prompted for it. This may have fixed the runtime error I was getting, and now yt-dlt works as expected, with less slug (??). I really don't know quite enough to know what I am doing ;-)

**Comment by low-batt** [comment]
This is line 1 of `yt-dlp`:
```text
#!/usr/bin/env python3
```

That is a [Shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)#Portability). It tells macOS to start `python3` and have it run `yt-dlp` which is implemented as a python script.

I'm pretty sure macOS does not include `python3`. Apparently something you just installed also installed `python3` and the `yt-dlp` can now start and run. 

One of the reasons `yt-dlp_macos` is so big and slow to unpack is that it includes the needed dependencies like `python3`.

**Comment by ksdavidc** [comment]
Again, a notably clear and comprehensive (and accurate) explainer. Thank you, that jibes. Open Video Downloader is a yt-dlp gui which I now have noticed, as you suggest, ["pre-installs" python](https://jely2002.github.io/youtube-dl-gui/#:~:text=%2D%20Python%20(comes%20pre%2Dinstalled)), whatever that means.

**Comment by low-batt** [comment]
I normally try and be detailed and clear in responses, but you caught me intentionally being vague. My "something you just installed" statement was due to me having researched Open Video Downloader and coming away confused about the exact statement you referenced, the ["pre-installs" python](https://jely2002.github.io/youtube-dl-gui/#:~:text=%2D%20Python%20(comes%20pre%2Dinstalled)) statement. The actual statement is this:
> MacOS:
> - Python (comes pre-installed)

Looking at the [Releases](https://github.com/jely2002/youtube-dl-gui) section of that repository shows the latest release is from Nov 13, **2021**.

Ancient by computer standards.

Back then `python` (not `python3`) came pre-installed in macOS.  It was removed in macOS 12.3 as noted in the [releases notes](https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python), which would have been after the last release of Open Video Downloader (from that repository).

I think that statement was referring to the `python` that used to come with macOS.

The python language was changed for version 3 of python in ways that could break scripts. To allow users to run older python scripts as well as new scripts the convention was to name the executable of the new version of python `python3`. Way more than you want to know about this can be found [here](https://peps.python.org/pep-0394/).

I was unclear on from where you obtained the Open Video Downloader app. I found Open Video Downloader [here](https://github.com/StefanLobbenmeier/youtube-dl-gui?tab=readme-ov-file#credit) which has this note:
> Big thanks to the original author [jely2002](https://github.com/jely2002), who unfortunately stopped maintaining the original repository: https://github.com/jely2002/youtube-dl-gui

This repositories's latest release is from Feb 18, 2025.

Anyway, I was being vague because that statement about python confused me was well. I now _think_ it was the Xcode that installed `python3`, although I could not find an authoritative statement from Apple on that.

**Comment by ksdavidc** [comment]
Oh, well now, we are really getting in the weeds:

- I originally downloaded Open Video Downloader (OVD) from the [Jelle Glebbeek](https://jelleglebbeek.com/) version found [here](https://jely2002.github.io/youtube-dl-gui/). I am not sure how I got directed there, i.e. via google, or another site. But, this is the old version 2.4, not the [StefanLobbenmeier](https://github.com/StefanLobbenmeier/youtube-dl-gui?tab=readme-ov-file#credit) version you mentioned. I didn't realize there was a newer version, and realized I should upgrade.
- To be precise I loaded Xcode Command Tools (XCT) not the full Xcode as it might have seemed above.
- So I uninstalled XCT, and reinstalled the new OVD. I received no prompt to install XCT. It seemed to be able to download videos fine.
-  Then in IIna, when I tried to upload a url this time, as expected, as there was no XCT, it bombed. 
- But putting in the correct yt-dlt folder and trying to get a url produced confirmation the XCT prompt had come from IIna, not OVD. I must have misidentified the source the first time.
- At the same time I confirmed that prior to XCT, python3 was not working, and directly after it was. I wasn't able to find direct evidence from Apple that this was generallly the case, but see: [When you install Xcode Command Line Tools, Apple includes Python 3.9.6.](https://mac.install.guide/python/system-default#:~:text=Check%20if%20a%20system%20Python%20is%20installed,-You%20can%20check&text=Try%20python3%20%2D%2Dversion%20and,with%20Xcode%20Command%20Line%20Tools.)
- My interpretation of the statement from OVD was a red herring after all.

**Comment by 0dr0colossus** [comment]
@low-batt I've been having this same problem as the OP for quite some time now, and thankfully fixing the custom path resolved my issue. I had to disable the Online Media plugin though — is that plugin even necessary, or is it busted? Apparently enabling it would *disable* my ability to change the custom path, but IINA wasn't able to stream youtube videos with the plugin's settings, and I had to change the path because I switched from Intel to an M-chip Mac about a couple years ago, so the old Intel path was still what IINA was using.

Basically I had to look up what the Online Media plugin even was — didn't even realize I had it — and the summary seems redundant, maybe even obstructive? I could very well be missing or misunderstanding context though.
